docs/rfcs/component-universal-telemetry.md

```diff
@@ -0,0 +1,119 @@
+# Auto-Instrumented Component Telemetry
+
+## Motivation
+
+The collector should be observable and this must naturally include observability of its pipeline components. It is understood that each _type_ (`filelog`, `batch`, etc) of component may emit telemetry describing its internal workings, and that these internally derived signals may vary greatly based on the concerns and maturity of each component. Naturally though, the collector should also describe the behavior of components using broadly normalized telemetry. A major challenge in pursuit is that there must be a clear mechanism by which such telemetry can be automatically captured. Therefore, this RFC is first and foremost a proposal for a _mechanism_. Then, based on what _can_ be captured by this mechanism, the RFC describes specific metrics, spans, and logs which can be broadly normalized.
+
+## Goals
+
+1. Articulate a mechanism which enables us to _automatically_ capture telemetry from _all pipeline components_.
+2. Define attributes that are (A) specific enough to describe individual component [_instances_](https://github.com/open-telemetry/opentelemetry-collector/issues/10534) and (B) consistent enough for correlation across signals.
+3. Define specific metrics for each kind of pipeline component.
+4. Define specific spans for processors and connectors.
+5. Define specific logs for all kinds of pipeline component.
+
+### Mechanism
+
+The mechanism of telemetry capture should be _external_ to components. Specifically, we should observe telemetry at each point where a component passes data to another component, and, at each point where a component consumes data from another component. In terms of the component graph, this means that every _edge_ in the graph will have two layers of instrumentation - one for the producing component and one for the consuming component. Importantly, each layer generates telemetry which is ascribed to a single component instance, so by having two layers per edge we can describe both sides of each handoff independently. In the case of processors and connectors, the appropriate layers can act in concert (e.g. record the start and end of a span).
+
+### Attributes
+
+All signals should use the following attributes:
+
+#### Receivers
+
+- `otel.component.kind`: `receiver`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs`, `metrics`, `traces`, **OR `ALL`**
+
+#### Processors
+
+- `otel.component.kind`: `processor`
+- `otel.component.id`: The component ID
+- `otel.pipeline.id`: The pipeline ID, **OR `ALL`**
+- `otel.signal`: `logs`, `metrics`, `traces`, **OR `ALL`**
+
+#### Exporters
+
+- `otel.component.kind`: `exporter`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs`, `metrics` `traces`, **OR `ALL`**
+
+#### Connectors
+
+- `otel.component.kind`: `connector`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs->logs`, `logs->metrics`, `logs->traces`, `metrics->logs`, `metrics->metrics`, etc, **OR `ALL`**
```

Comment:
If we annotate the incoming and outgoing counts with separate attributes, then connectors such as `count` or `otlpjson` will report a single data stream that aggregates across multiple instances. I believe this could lead to problems very similar to the one described [here](https://github.com/open-telemetry/opentelemetry-collector/issues/10908), where users need to be able to differentiate between instances in order to understand a problem. I think this is fundamentally a question of whether or not we should represent component _instances_ to users, or whether we're trying to aggregate and present them in some unified way. My opinion is that we have tried to gloss over this distinction for a long time but the nuances matter in some cases and we're not doing users a favor by hiding the distinction.

Reply:
I'm resolving this comment, but please reopen if you think the justification is not sufficient.